cartography/intel/azure/permission_relationships.py

```diff
@@ -0,0 +1,466 @@
+    """
+    # Expected principal types for validation
+    expected_types = {"User", "Group", "ServicePrincipal"}
+
+    mappings_by_type: dict[str, list[dict]] = {}
+
+    for mapping in principal_mappings:
+        assignment_principal_type = mapping["principal_type"]
+
+        # Validate principal type, no silent failure in case of change in expected principal types from MS
+        if assignment_principal_type not in expected_types:
+            logger.warning(
+                f"Unknown principal type '{assignment_principal_type}' encountered - skipping permission relationships sync for this principal type."
+            )
+            continue
+
+        # Add 'Entra' prefix to match Entra node labels
+        entra_principal_type = f"Entra{assignment_principal_type}"
+
+        if entra_principal_type not in mappings_by_type:
+            mappings_by_type[entra_principal_type] = []
+        mappings_by_type[entra_principal_type].append(mapping)
+
+    return mappings_by_type
+
+
+@timeit
+def load_principal_mappings(
+    neo4j_session: neo4j.Session,
+    mappings_by_type: dict[str, list[dict]],
+    node_label: str,
+    relationship_name: str,
+    update_tag: int,
+    subscription_id: str,
+) -> None:
+    if not mappings_by_type:
+        return
+
+    # Iterate over each principal type
+    principal_types = ["EntraUser", "EntraGroup", "EntraServicePrincipal"]
+
+    for principal_type in principal_types:
+        type_mappings = mappings_by_type.get(principal_type, [])
+
+        if not type_mappings:
+            continue
+
+        # Create MatchLink schema with dynamic attributes
+        matchlink_schema = AzurePermissionMatchLink(
+            source_node_label=principal_type,
+            target_node_label=node_label,
+            rel_label=relationship_name,
+        )
+
+        logger.info(
+            f"Loading {len(type_mappings)} {relationship_name} relationships for {principal_type} -> {node_label}"
+        )
+
+        load_matchlinks(
+            neo4j_session,
+            matchlink_schema,
+            type_mappings,
+            lastupdated=update_tag,
+            _sub_resource_label="AzureSubscription",
+            _sub_resource_id=subscription_id,
+        )
+
+
+@timeit
+def cleanup_rpr(
+    neo4j_session: neo4j.Session,
+    node_label: str,
+    relationship_name: str,
+    update_tag: int,
+    subscription_id: str,
+) -> None:
+    logger.info(
+        "Cleaning up relationship '%s' for node label '%s'",
+        relationship_name,
+        node_label,
+    )
+
+    # Clean up for each principal type
+    principal_types = ["EntraUser", "EntraGroup", "EntraServicePrincipal"]
+    for principal_type in principal_types:
+        # Create MatchLink schema with dynamic attributes
+        matchlink_schema = AzurePermissionMatchLink(
+            source_node_label=principal_type,
+            target_node_label=node_label,
+            rel_label=relationship_name,
+        )
```

Comment:
No need to do this dynamic creation of schema objects with a for loop. for the sake of clairty here, we should be explicit instead even if it is repeating ourselves.

Reply:
As discussed, because rel_label is user specified and can be anything, it is ok to let all these fields be dynamic